src/jtag3.c

```diff
@@ -2513,6 +2491,399 @@ static unsigned int jtag3_memaddr(const PROGRAMMER *pgm, const AVRPART *p, const
+    pmsg_error("no flash memory for part %s\n", p->desc);
+    return LIBAVRDUDE_GENERAL_FAILURE;
+  }
+
+  // An erase is triggered by an access to the hi-byte
+  paddr = m->offset | 0x01;
+
+  cmd[0] = XPRG_CMD_ERASE;
+  cmd[1] = XPRG_ERASE_CHIP;
+  u32_to_b4_big_endian((cmd+2), paddr);
+
+  if ((status = jtag3_command_tpi(pgm, cmd, len, &resp, "chip erase")) < 0)
+    return -1;
+  free(resp);
+  return 0;
+}
+
+static int jtag3_open_tpi(PROGRAMMER *pgm, const char *port) {
+  pmsg_notice2("jtag3_open_tpi()\n");
+
+  if (jtag3_open_common(pgm, port) < 0)
+    return -1;
+  return 0;
+}
+
+void jtag3_close_tpi(PROGRAMMER *pgm) {
+  pmsg_notice2("jtag3_close_tpi() is empty. No action necessary.\n");
+}
+
+static int jtag3_paged_load_tpi(const PROGRAMMER *pgm, const AVRPART *p,
+                                const AVRMEM *m, unsigned int page_size,
+                                unsigned int addr, unsigned int n_bytes) {
+  unsigned int block_size = 0;
+  unsigned int maxaddr = addr + n_bytes;
+  unsigned char cmd[8];
+  unsigned char *resp;
+  int status;
+  long otimeout = serial_recv_timeout;
+
+  msg_notice2("\n");
+  pmsg_notice2("jtag3_paged_load_tpi(.., %s, %d, 0x%04x, %d)\n",
+               m->desc, page_size, addr, n_bytes);
+
+  if(m->offset)
+    msg_notice2("          mapped to address: 0x%04x\n", (addr+m->offset));
+
+  cmd[0] = XPRG_CMD_READ_MEM;
+  cmd[1] = tpi_get_memtype(m);
+
+  if(m->blocksize > page_size)
+    page_size = m->blocksize;
+
+  serial_recv_timeout = 100;
+  for (; addr < maxaddr; addr += page_size) {
+    if ((maxaddr - addr) < page_size)
+      block_size = maxaddr - addr;
+    else
+      block_size = page_size;
+    pmsg_debug("jtag3_paged_load_tpi(): "
+               "block_size at addr 0x%x is %d\n", addr, block_size);
+
+    u32_to_b4_big_endian((cmd+2), addr + m->offset);  // Address
+    u16_to_b2_big_endian((cmd+6), block_size);        // Size
+
+    if ((status = jtag3_command_tpi(pgm, cmd, 8, &resp, "read memory")) < 0)
+      return -1;
+
+    if (resp[1] != XPRG_ERR_OK ||
+        status < block_size + 2) {
+      pmsg_error("wrong/short reply to read memory command\n");
+      serial_recv_timeout = otimeout;
+      free(resp);
+      return -1;
+    }
+    memcpy(m->buf + addr, resp + 2, status - 2);
+    free(resp);
+  }
+  serial_recv_timeout = otimeout;
+
+  return n_bytes;
+}
+
+static int jtag3_paged_write_tpi(const PROGRAMMER *pgm, const AVRPART *p,
+                                 const AVRMEM *m, unsigned int page_size,
+                                 unsigned int addr, unsigned int n_bytes) {
+  unsigned int block_size;
+  unsigned int maxaddr = addr + n_bytes;
+  unsigned char *cmd;
+  unsigned char *resp;
+  int status;
+  long otimeout = serial_recv_timeout;
+
+  msg_notice2("\n");
+  pmsg_notice2("jtag3_paged_write_tpi(.., %s, %d, 0x%04x, %d)\n", m->desc, page_size, addr, n_bytes);
+
+  if(m->offset)
+    msg_notice2("          mapped to address: 0x%04x\n", (addr+m->offset));
```

Comment:
For indented message you can use `imsg_notice2` instead (or any `imsg` function)

Reply:
Nice :-)